api/types/databaseobject/databaseobject.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package databaseobject
+
+import (
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/defaults"
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// NewDatabaseObject creates a new dbobjectv1.DatabaseObject.
+func NewDatabaseObject(name string, spec *dbobjectv1.DatabaseObjectSpec) (*dbobjectv1.DatabaseObject, error) {
+	return NewDatabaseObjectWithLabels(name, nil, spec)
+}
+
+// NewDatabaseObjectWithLabels creates a new dbobjectv1.DatabaseObject with specified labels.
+func NewDatabaseObjectWithLabels(name string, labels map[string]string, spec *dbobjectv1.DatabaseObjectSpec) (*dbobjectv1.DatabaseObject, error) {
+	databaseObject := &dbobjectv1.DatabaseObject{
+		Kind:    types.KindDatabaseObject,
+		Version: types.V1,
+		Metadata: &headerv1.Metadata{
+			Name:      name,
+			Namespace: defaults.Namespace,
+			Labels:    labels,
+		},
+		Spec: spec,
+	}
+
+	err := ValidateDatabaseObject(databaseObject)
```

Comment:
I'm of the strong opinion that validation is a storage responsibility. This should be gated by storage during Create/Update/Upsert and not whenever a new resource is initialized.

This pattern bakes validation into clients and into multiple layers, diffusing responsibility and making validation changes difficult to do.

* https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#resource-validation

Reply:
The linked sections says:

> Resource validation
The strictest validation of a resource should be performed prior to write operations. Any resource persisted in the backend should be guaranteed to be valid. Read operations should not perform resource validations, doing so could prevent a resource being read if validations are modified to be more restrictive after a resource had already been written. This allows for resilient reads which ensure that any resource stored is always allowed to be retrieved from the backend.

It does not say that no validation should occur outside of storage. It does the opposite, really: using the word "strictest" suggest some others are less strict.

More generally, performing validation on the write path doesn't help with usage of ephemeral objects. It is a common pattern to validate an object on creation, i.e. in `NewFoo()` function for type `Foo`. This only makes sense: you want rapid feedback if you are trying to create an invalid object.

Now, I can see an argument for creating two tiers of validation, one more strict than the other, and only using the strict form on writes. But that sounds like unnecessary complexity, as you now have to carefully maintain two distinct sets of requirements for no clear benefit.

Reply:
The decider of what "valid" means is typically the backend. Pushing validation to New functions can make clients become out of sync with the backends and over-validate. It can also cause problems when reading stored values that "New" considers invalid (but should be treated resiliently instead).

Agree to disagree.